### PR TITLE
20190529_async-promise-improvement: Async promise finish conditions

### PR DESCRIPTION
**Pull request by yekhlakov**

### CODE REVIEW

File: src/Jobs/Promise.php [hunk 1]
```diff
@@ -230,104 +261,126 @@ protected function deleteRaw()
...
-                    }
-                }
-            }
-        } catch (PromiseNotFoundException $e) {
-            DB::rollBack();
-        } catch (\Exception $e) {
-            DB::rollBack();
-            throw $e;
+        if ($nextJob && $this->runNextJob($nextJob)) {
+            $nextJob->dispatchJob($nextJob);
```
Comment (darkdarin) [inline src/Jobs/Promise.php]: $this

File: src/Jobs/Promise.php [hunk 2]
```diff
@@ -230,104 +261,126 @@ protected function deleteRaw()
...
+        if ($this->promise_type === self::PROMISE_TYPE_SYNC) {
+            return $jobStatus === MayPromised::JOB_STATUS_ERROR;
+        }
 
-            if (!$promise) {
-                throw new PromiseNotFoundException('Promise #'. $job->getPromiseId() .' not found');
-            }
+        // Для асинхронного - все зависит от флагов.
+        if ($jobStatus === MayPromised::JOB_STATUS_SUCCESS) {
+            return !$this->promise_finish_on_first_success;
```
Comment (darkdarin) [inline src/Jobs/Promise.php]: inverse

File: src/Jobs/Promise.php [hunk 3]
```diff
@@ -230,104 +261,126 @@ protected function deleteRaw()
...
+        }
 
-            // если такого запроса нет - игнорируем
-            if (!isset($promise->promise_jobs[$job->getUniqueId()])) {
-                throw new PromiseNotFoundException('Job #'. $job->getUniqueId() .' in promise ' . $job->getPromiseId() . ' not found');
-            }
+        // Если мы здесь, то джоб либо завершен с ошибкой, либо в неизвестном статусе
+        // Неизвестный статус приравниваем к ошибке
+        return !$this->promise_finish_on_first_error;
```
Comment (darkdarin) [inline src/Jobs/Promise.php]: inverse

File: src/Jobs/Promise.php (hunk 1 above)
Comment (yekhlakov) [reply src/Jobs/Promise.php]: done

File: src/Jobs/Promise.php (hunk 2 above)
Comment (yekhlakov) [reply src/Jobs/Promise.php]: done

File: src/Jobs/Promise.php (hunk 3 above)
Comment (yekhlakov) [reply src/Jobs/Promise.php]: done

**Pull request merged**